pretty.c

```diff
@@ -0,0 +1,876 @@
+{
+	printf("%s" L_LAST "%s\n", prefix, json_object_get_string(obj));
+}
+
+static void print_properties(struct json_object *obj, const char *prefix)
+{
+	printf("%s" L_LAST "Properties\n", prefix);
+
+	struct json_object_iter iter;
+	json_object_object_foreachC(obj, iter) {
+		bool last = !iter.entry->next;
+		const char *prop_name = iter.key;
+		struct json_object *prop_obj = iter.val;
+
+		char sub_prefix[strlen(prefix) + 2 * strlen(L_VAL) + 1];
+		snprintf(sub_prefix, sizeof(sub_prefix), "%s" L_GAP "%s",
+			prefix, last ? L_GAP : L_LINE);
+
+		uint32_t flags = get_object_object_uint64(prop_obj, "flags");
+		uint32_t type = flags & (DRM_MODE_PROP_LEGACY_TYPE |
+			DRM_MODE_PROP_EXTENDED_TYPE);
+		bool atomic = flags & DRM_MODE_PROP_ATOMIC;
+		bool immutable = flags & DRM_MODE_PROP_IMMUTABLE;
+
+		printf("%s" L_GAP "%s\"%s\"", prefix, last ? L_LAST : L_VAL, prop_name);
+		if (atomic && immutable)
+			printf(" (atomic, immutable)");
+		else if (atomic)
+			printf(" (atomic)");
+		else if (immutable)
+			printf(" (immutable)");
+
+		printf(": ");
+
+		uint64_t raw_val = get_object_object_uint64(prop_obj, "raw_value");
+		struct json_object *spec_obj = json_object_object_get(prop_obj, "spec");
+		struct json_object *val_obj = json_object_object_get(prop_obj, "value");
+		struct json_object *data_obj = json_object_object_get(prop_obj, "data");
+		bool first;
+		switch (type) {
+		case DRM_MODE_PROP_RANGE:;
+			uint64_t min = get_object_object_uint64(spec_obj, "min");
+			uint64_t max = get_object_object_uint64(spec_obj, "max");
+			const char *min_str = u64_str(min);
+			const char *max_str = u64_str(max);
+
+			if (min_str)
+				printf("range [%s, ", min_str);
+			else
+				printf("range [%"PRIu64", ", min);
+
+			if (max_str)
+				printf("%s]", max_str);
+			else
+				printf("%"PRIu64"]", max);
+
+			if (data_obj != NULL)
+				printf(" = %"PRIu64"\n", get_object_uint64(data_obj));
+			else
+				printf(" = %"PRIu64"\n", get_object_uint64(val_obj));
+			break;
+		case DRM_MODE_PROP_ENUM:
+			printf("enum {");
+			const char *val_name = NULL;
+			first = true;
+			for (size_t j = 0; j < json_object_array_length(spec_obj); ++j) {
+				struct json_object *item_obj =
+					json_object_array_get_idx(spec_obj, j);
+				const char *item_name =
+					get_object_object_string(item_obj, "name");
+				uint64_t item_value =
+					get_object_object_uint64(item_obj, "value");
+
+				if (raw_val == item_value) {
+					val_name = item_name;
+				}
+
+				printf("%s%s", first ? "" : ", ", item_name);
+				first = false;
+			}
+			printf("} = ");
+
+			if (val_name) {
+				printf("%s\n", val_name);
+			} else {
+				printf("invalid (%"PRIu64")\n", raw_val);
+			}
+			break;
+		case DRM_MODE_PROP_BLOB:;
+			printf("blob = %" PRIu64 "\n", raw_val);
+			if (strcmp(prop_name, "IN_FORMATS") == 0)
+				print_in_formats(data_obj, sub_prefix);
+			else if (strcmp(prop_name, "MODE_ID") == 0)
+				print_mode_id(data_obj, sub_prefix);
```

Comment:
This crashes when MODE_ID == 0, i.e. data_obj == NULL.
But I think that would apply to the other properties here too.